src/Conversion/ONNXToKrnl/Tensor/DepthToSpace.cpp

```diff
@@ -0,0 +1,185 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===------------ DepthToSpace.cpp - Lowering DepthToSpace Op -------------===//
+//
+// Copyright 2019 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers the ONNX DepthToSpace Operator to Krnl dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+#include "src/Dialect/Krnl/KrnlHelper.hpp"
+#include "src/Dialect/ONNX/ONNXShapeHelper.hpp"
+#include "llvm/Support/Debug.h"
+
+using namespace mlir;
+using llvm::dbgs;
+
+#define DEBUG_TYPE "depth_to_space_onnx_to_krnl"
+
+struct ONNXDepthToSpaceOpLowering : public ConversionPattern {
+  ONNXDepthToSpaceOpLowering(MLIRContext *ctx)
+      : ConversionPattern(
+            mlir::ONNXDepthToSpaceOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    auto spaceToDepthOp = dyn_cast_or_null<ONNXDepthToSpaceOp>(op);
+    assert(spaceToDepthOp && "Expecting op to have type ONNXDepthToSpaceOp");
+
+    // Ensure we can compute the operator output shape.
+    ONNXDepthToSpaceOpShapeHelper shapeHelper(&spaceToDepthOp, &rewriter,
+        getDenseElementAttributeFromKrnlValue,
+        loadDenseElementArrayValueAtIndex);
+    ONNXDepthToSpaceOpAdaptor operandAdaptor(operands);
+    LogicalResult shapeComputed = shapeHelper.computeShape(operandAdaptor);
+    (void)shapeComputed;
+    assert(succeeded(shapeComputed) && "Could not compute output shape");
+
+    Value input = spaceToDepthOp.input();
+    int64_t bs = spaceToDepthOp.blocksize();
+    Location loc = spaceToDepthOp.getLoc();
+    StringRef mode = spaceToDepthOp.mode();
+
+    // Compute the new dimensions.
+    MemRefBoundsIndexCapture bounds(input);
+    assert(bounds.getRank() == 4 && "Input tensor should have rank equal to 4");
+
+    DimIndexExpr B(bounds.getDim(0));
+    DimIndexExpr C(bounds.getDim(1));
+    DimIndexExpr H(bounds.getDim(2));
+    DimIndexExpr W(bounds.getDim(3));
+    DimIndexExpr newC = create<SignedFloorDivIOp>(C, (bs * bs), rewriter, loc);
+    DimIndexExpr newH = create<MulIOp>(H, bs, rewriter, loc);
+    DimIndexExpr newW = create<MulIOp>(W, bs, rewriter, loc);
+
```

Comment:
IndexExpr supports Mul, FloorDiv, so we can write, say, `newC = C.floorDiv(bs*bs)`, `newH = H * bs`, `newW = W*bs` for brevity.